### ISSUE TITLE
[BUG] Query generation in 0.1.6 differs to 0.1.5

### ISSUE THREAD

**Issue by anthonynorth** [ropensci/osmdata#240]
Generated queries in v0.1.6 differ from v0.1.5; it appears that repeated `osmdata::add_osm_feature()` calls are `or` separated now?

I've traced the change in behviour to `opq_string_intern`: [v0.1.6](https://github.com/ropensci/osmdata/blame/v0.1.6/R/opq.R#L457-L460) & [0.1.5](https://github.com/ropensci/osmdata/blame/v0.1.5/R/opq.R#L393)

The previous behaviour collapsed multiple features into a single string and new behaviour doesn't. The result of this is that _node_, _way_ and _relation_ blocks are applied to each feature separately (relevant code here https://github.com/ropensci/osmdata/blob/main/R/opq.R#L484-L492). 

You can observe the difference in behaviour in the two reprex below. I can reproduce v0.1.5 behaviour in v0.1.6 by _patching_ the opq object prior to `osmdata::osmdata_sc` with:

```r
# revert to v0.1.5 behaviour
patch_features <- function(opq) {
  opq$features <- paste0(opq$features, collapse = "")
  opq
}
```

## Reprex: v0.1.6

``` r
bbox <- sf::st_bbox(
  c(xmin = 146.81582, ymin = -19.26773, xmax = 146.82536, ymax = -19.25866),
  crs = 4326
)

highway_types <- c(
  "motorway", "primary", "motorway_link", "primary_link", "secondary",
  "tertiary", "secondary_link", "tertiary_link", "residential", "living_street",
  "unclassified", "service", "road", "trunk", "trunk_link"
)

osmquery <- osmdata::opq(bbox) |>
  osmdata::add_osm_feature(key = "highway", value = c(highway_types)) |>
  osmdata::add_osm_feature(key = "area", value = "!yes")

# raw query
print(osmdata:::opq_string_intern(osmquery))
#> [1] "[out:xml][timeout:25];\n(\n node  [\"highway\"~\"^(motorway|primary|motorway_link|primary_link|secondary|tertiary|secondary_link|tertiary_link|residential|living_street|unclassified|service|road|trunk|trunk_link)$\"] (-19.26773,146.81582,-19.25866,146.82536);\n way  [\"highway\"~\"^(motorway|primary|motorway_link|primary_link|secondary|tertiary|secondary_link|tertiary_link|residential|living_street|unclassified|service|road|trunk|trunk_link)$\"] (-19.26773,146.81582,-19.25866,146.82536);\n relation  [\"highway\"~\"^(motorway|primary|motorway_link|primary_link|secondary|tertiary|secondary_link|tertiary_link|residential|living_street|unclassified|service|road|trunk|trunk_link)$\"] (-19.26773,146.81582,-19.25866,146.82536);\n\n node  [\"area\"!=\"yes\"] (-19.26773,146.81582,-19.25866,146.82536);\n way  [\"area\"!=\"yes\"] (-19.26773,146.81582,-19.25866,146.82536);\n relation  [\"area\"!=\"yes\"] (-19.26773,146.81582,-19.25866,146.82536);\n\n);\n(._;>;);\nout body;"

osmdata::osmdata_sc(osmquery)
#> $nodes
#> # A tibble: 522 x 3
#>    vertex_  key     value          
#>    <chr>    <chr>   <chr>          
#>  1 31274560 highway traffic_signals
#>  2 31274565 highway traffic_signals
#>  3 31274902 highway traffic_signals
#>  4 31275711 highway traffic_signals
#>  5 31452861 highway traffic_signals
#>  6 31452865 highway traffic_signals
#>  7 31452974 highway traffic_signals
#>  8 31475029 highway traffic_signals
#>  9 31475102 highway traffic_signals
#> 10 31485961 railway level_crossing 
#> # ... with 512 more rows
#> 
#> $relation_members
#> # A tibble: 1,478 x 4
#>    relation_ member     type  role   
#>    <chr>     <chr>      <chr> <chr>  
#>  1 2270945   170745952  way   "outer"
#>  2 2270945   170745951  way   "inner"
#>  3 2868305   2258936770 node  ""     
#>  4 2868305   2258936778 node  ""     
#>  5 2868305   216538283  way   ""     
#>  6 2868305   36962939   way   ""     
#>  7 2868305   182851432  way   ""     
#>  8 2868305   36962888   way   ""     
#>  9 2868305   200690222  way   ""     
#> 10 2868305   216538268  way   ""     
#> # ... with 1,468 more rows
#> 
#> $relation_properties
#> # A tibble: 98 x 3
#>    relation_ key      value           
#>    <chr>     <chr>    <chr>           
#>  1 2270945   leisure  park            
#>  2 2270945   name     Dean Street Park
#>  3 2270945   type     multipolygon    
#>  4 2868305   name     201             
#>  5 2868305   network  TAG             
#>  6 2868305   operator Sunbus          
#>  7 2868305   ref      201             
#>  8 2868305   route    bus             
#>  9 2868305   type     route           
#> 10 2868346   name     200             
#> # ... with 88 more rows
#> 
#> $object
#> # A tibble: 6,016 x 3
#>    object_ key         value                
#>    <chr>   <chr>       <chr>                
#>  1 4851599 highway     primary              
#>  2 4851599 junction    roundabout           
#>  3 4851599 surface     paved                
#>  4 4851601 highway     primary              
#>  5 4851601 name        Duckworth Street     
#>  6 4851601 oneway      yes                  
#>  7 4851601 source:name voice                
#>  8 4851601 surface     paved                
#>  9 4851613 bridge      yes                  
#> 10 4851613 bridge:name Charles Barton Bridge
#> # ... with 6,006 more rows
#> 
#> $object_link_edge
#> # A tibble: 9,570 x 3
#>    edge_      object_ native_
#>    <chr>      <chr>   <lgl>  
#>  1 Uq0LBBVoVA 4851599 TRUE   
#>  2 XJw6u0mfzC 4851599 TRUE   
#>  3 5BLa13JMQn 4851599 TRUE   
#>  4 CGj704FA2F 4851599 TRUE   
#>  5 i2tzOXTBSn 4851599 TRUE   
#>  6 5cth4bmJ7C 4851599 TRUE   
#>  7 SbdK1cNKvA 4851599 TRUE   
#>  8 hrfUJoqrce 4851599 TRUE   
#>  9 D3XCfkgl4r 4851599 TRUE   
#> 10 mXFEIOZkSw 4851599 TRUE   
#> # ... with 9,560 more rows
#> 
#> $edge
#> # A tibble: 9,570 x 3
#>    .vx0       .vx1       edge_     
#>    <chr>      <chr>      <chr>     
#>  1 31274426   1490103778 Uq0LBBVoVA
#>  2 1490103778 2102391162 XJw6u0mfzC
#>  3 2102391162 1490103786 5BLa13JMQn
#>  4 1490103786 31274451   CGj704FA2F
#>  5 31274451   1490103768 i2tzOXTBSn
#>  6 1490103768 31274452   5cth4bmJ7C
#>  7 31274452   1490103776 SbdK1cNKvA
#>  8 1490103776 31274453   hrfUJoqrce
#>  9 31274453   1490103769 D3XCfkgl4r
#> 10 1490103769 31274454   mXFEIOZkSw
#> # ... with 9,560 more rows
#> 
#> $vertex
#> # A tibble: 9,252 x 3
#>       x_    y_ vertex_ 
#>    <dbl> <dbl> <chr>   
#>  1  147. -19.3 31274426
#>  2  147. -19.3 31274451
#>  3  147. -19.3 31274452
#>  4  147. -19.3 31274453
#>  5  147. -19.3 31274454
#>  6  147. -19.3 31274455
#>  7  147. -19.3 31274456
#>  8  147. -19.3 31274457
#>  9  147. -19.3 31274472
#> 10  147. -19.3 31274552
#> # ... with 9,242 more rows
#> 
#> $meta
#> # A tibble: 1 x 5
#>   proj  ctime                       OSM_version overpass_version    bbox        
#>   <chr> <chr>                       <chr>       <chr>               <chr>       
#> 1 <NA>  [ Fri 6 Aug 2021 07:11:50 ] 0.6         Overpass API 0.7.5~ -19.26773,1~
#> 
#> attr(,"join_ramp")
#> [1] "nodes"               "relation_members"    "relation_properties"
#> [4] "object"              "object_link_edge"    "edge"               
#> [7] "vertex"             
#> attr(,"class")
#> [1] "SC"         "sc"         "osmdata_sc"
```

<sup>Created on 2021-08-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessionInfo()
#> R version 4.1.0 (2021-05-18)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19042)
#> 
#> Matrix products: default
#> 
#> locale:
#> [1] LC_COLLATE=English_Australia.1252  LC_CTYPE=English_Australia.1252   
#> [3] LC_MONETARY=English_Australia.1252 LC_NUMERIC=C                      
#> [5] LC_TIME=English_Australia.1252    
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> loaded via a namespace (and not attached):
#>  [1] Rcpp_1.0.7         pillar_1.6.2       compiler_4.1.0     highr_0.9         
#>  [5] class_7.3-19       tools_4.1.0        digest_0.6.27      lubridate_1.7.10  
#>  [9] jsonlite_1.7.2     evaluate_0.14      lifecycle_1.0.0    tibble_3.1.3      
#> [13] lattice_0.20-44    pkgconfig_2.0.3    rlang_0.4.11       reprex_2.0.1      
#> [17] cli_3.0.1          DBI_1.1.1          curl_4.3.2         yaml_2.2.1        
#> [21] xfun_0.25          e1071_1.7-8        xml2_1.3.2         httr_1.4.2        
#> [25] withr_2.4.2        styler_1.5.1       stringr_1.4.0      knitr_1.33        
#> [29] generics_0.1.0     fs_1.5.0           vctrs_0.3.8        classInt_0.4-3    
#> [33] grid_4.1.0         glue_1.4.2         R6_2.5.1           sf_1.0-2          
#> [37] fansi_0.5.0        rmarkdown_2.10     sp_1.4-5           osmdata_0.1.6     
#> [41] purrr_0.3.4        magrittr_2.0.1     backports_1.2.1    ellipsis_0.3.2    
#> [45] htmltools_0.5.1.1  units_0.7-2        rvest_1.0.1        utf8_1.2.2        
#> [49] KernSmooth_2.23-20 stringi_1.7.3      proxy_0.4-26       crayon_1.4.1
```

</details>

## Reprex: v0.1.5

``` r
bbox <- sf::st_bbox(
  c(xmin = 146.81582, ymin = -19.26773, xmax = 146.82536, ymax = -19.25866),
  crs = 4326
)

highway_types <- c(
  "motorway", "primary", "motorway_link", "primary_link", "secondary",
  "tertiary", "secondary_link", "tertiary_link", "residential", "living_street",
  "unclassified", "service", "road", "trunk", "trunk_link"
)

osmquery <- osmdata::opq(bbox) |>
  osmdata::add_osm_feature(key = "highway", value = c(highway_types)) |>
  osmdata::add_osm_feature(key = "area", value = "!yes")

# raw query
print(osmdata:::opq_string_intern(osmquery))
#> [1] "[out:xml][timeout:25];\n(\n node  [\"highway\"~\"^(motorway|primary|motorway_link|primary_link|secondary|tertiary|secondary_link|tertiary_link|residential|living_street|unclassified|service|road|trunk|trunk_link)$\"] [\"area\"!=\"yes\"] (-19.26773,146.81582,-19.25866,146.82536);\n way  [\"highway\"~\"^(motorway|primary|motorway_link|primary_link|secondary|tertiary|secondary_link|tertiary_link|residential|living_street|unclassified|service|road|trunk|trunk_link)$\"] [\"area\"!=\"yes\"] (-19.26773,146.81582,-19.25866,146.82536);\n relation  [\"highway\"~\"^(motorway|primary|motorway_link|primary_link|secondary|tertiary|secondary_link|tertiary_link|residential|living_street|unclassified|service|road|trunk|trunk_link)$\"] [\"area\"!=\"yes\"] (-19.26773,146.81582,-19.25866,146.82536);\n\n);\n(._;>;);\nout body;"

osmdata::osmdata_sc(osmquery)
#> $nodes
#> # A tibble: 33 x 3
#>    vertex_    key     value          
#>    <chr>      <chr>   <chr>          
#>  1 32443554   railway level_crossing 
#>  2 32443555   railway level_crossing 
#>  3 32462732   highway turning_circle 
#>  4 60447449   highway traffic_signals
#>  5 60447450   highway traffic_signals
#>  6 429694469  highway traffic_signals
#>  7 440981503  highway traffic_signals
#>  8 565331809  highway traffic_signals
#>  9 566522042  highway traffic_signals
#> 10 1490833981 highway traffic_signals
#> # ... with 23 more rows
#> 
#> $relation_members
#> # A tibble: 0 x 4
#> # ... with 4 variables: relation_ <chr>, member <chr>, type <chr>, role <chr>
#> 
#> $relation_properties
#> # A tibble: 0 x 3
#> # ... with 3 variables: relation_ <chr>, key <chr>, value <chr>
#> 
#> $object
#> # A tibble: 678 x 3
#>    object_ key         value                
#>    <chr>   <chr>       <chr>                
#>  1 4885468 bridge      yes                  
#>  2 4885468 bridge:name George Roberts Bridge
#>  3 4885468 highway     secondary            
#>  4 4885468 layer       1                    
#>  5 4885468 name        George Roberts Bridge
#>  6 4885468 network     S                    
#>  7 4885468 oneway      yes                  
#>  8 4885468 ref         16                   
#>  9 4885468 surface     paved                
#> 10 4934039 highway     residential          
#> # ... with 668 more rows
#> 
#> $object_link_edge
#> # A tibble: 478 x 3
#>    edge_      object_ native_
#>    <chr>      <chr>   <lgl>  
#>  1 zetXci80ke 4885468 TRUE   
#>  2 kwFj1MFyjv 4885468 TRUE   
#>  3 fvvD6HQZXO 4885468 TRUE   
#>  4 iRFIQU0LlC 4885468 TRUE   
#>  5 DECFiBRlhg 4885468 TRUE   
#>  6 7YpAIXmFMP 4934039 TRUE   
#>  7 HmhfVjpek7 4934039 TRUE   
#>  8 UThdWkNxjV 4934039 TRUE   
#>  9 JlLIgwDoBR 4934039 TRUE   
#> 10 OW7s82ivmV 4934039 TRUE   
#> # ... with 468 more rows
#> 
#> $edge
#> # A tibble: 478 x 3
#>    .vx0       .vx1       edge_     
#>    <chr>      <chr>      <chr>     
#>  1 60447729   7069479404 zetXci80ke
#>  2 7069479404 7069479400 kwFj1MFyjv
#>  3 7069479400 7069479403 fvvD6HQZXO
#>  4 7069479403 7069479402 iRFIQU0LlC
#>  5 7069479402 31676235   DECFiBRlhg
#>  6 5541023670 32462690   7YpAIXmFMP
#>  7 32462690   426266687  HmhfVjpek7
#>  8 426266687  3140203540 UThdWkNxjV
#>  9 3140203540 32443554   JlLIgwDoBR
#> 10 32443554   32462691   OW7s82ivmV
#> # ... with 468 more rows
#> 
#> $vertex
#> # A tibble: 417 x 3
#>       x_    y_ vertex_ 
#>    <dbl> <dbl> <chr>   
#>  1  147. -19.3 31676194
#>  2  147. -19.3 31676195
#>  3  147. -19.3 31676196
#>  4  147. -19.3 31676197
#>  5  147. -19.3 31676198
#>  6  147. -19.3 31676235
#>  7  147. -19.3 32443554
#>  8  147. -19.3 32443555
#>  9  147. -19.3 32444119
#> 10  147. -19.3 32444190
#> # ... with 407 more rows
#> 
#> $meta
#> # A tibble: 1 x 5
#>   proj  ctime                       OSM_version overpass_version    bbox        
#>   <chr> <chr>                       <chr>       <chr>               <chr>       
#> 1 <NA>  [ Fri 6 Aug 2021 07:41:12 ] 0.6         Overpass API 0.7.5~ -19.26773,1~
#> 
#> attr(,"join_ramp")
#> [1] "nodes"               "relation_members"    "relation_properties"
#> [4] "object"              "object_link_edge"    "edge"               
#> [7] "vertex"             
#> attr(,"class")
#> [1] "SC"         "sc"         "osmdata_sc"
```

<sup>Created on 2021-08-20 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessionInfo()
#> R version 4.1.0 (2021-05-18)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19042)
#> 
#> Matrix products: default
#> 
#> locale:
#> [1] LC_COLLATE=English_Australia.1252  LC_CTYPE=English_Australia.1252   
#> [3] LC_MONETARY=English_Australia.1252 LC_NUMERIC=C                      
#> [5] LC_TIME=English_Australia.1252    
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> loaded via a namespace (and not attached):
#>  [1] Rcpp_1.0.7         pillar_1.6.2       compiler_4.1.0     highr_0.9         
#>  [5] class_7.3-19       tools_4.1.0        digest_0.6.27      lubridate_1.7.10  
#>  [9] jsonlite_1.7.2     lattice_0.20-44    evaluate_0.14      lifecycle_1.0.0   
#> [13] tibble_3.1.3       pkgconfig_2.0.3    rlang_0.4.11       reprex_2.0.1      
#> [17] cli_3.0.1          DBI_1.1.1          curl_4.3.2         yaml_2.2.1        
#> [21] xfun_0.25          e1071_1.7-8        xml2_1.3.2         httr_1.4.2        
#> [25] withr_2.4.2        styler_1.5.1       stringr_1.4.0      dplyr_1.0.7       
#> [29] knitr_1.33         generics_0.1.0     fs_1.5.0           vctrs_0.3.8       
#> [33] tidyselect_1.1.1   classInt_0.4-3     grid_4.1.0         glue_1.4.2        
#> [37] sf_1.0-2           R6_2.5.1           fansi_0.5.0        rmarkdown_2.10    
#> [41] sp_1.4-5           osmdata_0.1.5      purrr_0.3.4        magrittr_2.0.1    
#> [45] backports_1.2.1    ellipsis_0.3.2     htmltools_0.5.1.1  units_0.7-2       
#> [49] rvest_1.0.1        assertthat_0.2.1   utf8_1.2.2         KernSmooth_2.23-20
#> [53] stringi_1.7.3      proxy_0.4-26       crayon_1.4.1
```

</details>

**Comment by mpadge** [comment]
Thanks @anthonynorth, and you were absolutely correct there. The above commit fixes and gives this:

``` r
library (osmdata)
#> Data (c) OpenStreetMap contributors, ODbL 1.0. https://www.openstreetmap.org/copyright
packageVersion ("osmdata")
#> [1] '0.1.6.2'

q0 <- opq(bbox = 'Kunming, China') |>
    add_osm_feature(key = 'natural', value = 'water') |>
    add_osm_feature(key = 'name:en', value = 'Dian', value_exact = FALSE)
opq_string (q0)
#> [1] "[out:xml][timeout:25];\n(\n 
#> node  [\"natural\"=\"water\"] [\"name:en\"~\"Dian\"] (24.388848,102.1697441,26.548485,103.6683522);\n 
#> way  [\"natural\"=\"water\"] [\"name:en\"~\"Dian\"] (24.388848,102.1697441,26.548485,103.6683522);\n 
#> relation  [\"natural\"=\"water\"] [\"name:en\"~\"Dian\"] (24.388848,102.1697441,26.548485,103.6683522);\n\n);
#> \n(._;>;);\nout body;"
length (q0$features)
#> [1] 1
x0 <- osmdata_sf (q0)
length (x0$osm_polygons)
#> [1] 3

q1 <- opq(bbox = 'Kunming, China') |>
    add_osm_features(features = c ("\"natural\"=\"water\"",
                                   "\"name:en\"=\"Dian\""))
length (q1$features)
#> [1] 2
opq_string (q1)
#> [1] "[out:xml][timeout:25];\n(\n 
#> node [\"natural\"=\"water\"] (24.388848,102.1697441,26.548485,103.6683522);\n 
#> way [\"natural\"=\"water\"] (24.388848,102.1697441,26.548485,103.6683522);\n 
#> relation [\"natural\"=\"water\"] (24.388848,102.1697441,26.548485,103.6683522);\n\n 
#> node [\"name:en\"=\"Dian\"] (24.388848,102.1697441,26.548485,103.6683522);\n 
#> way [\"name:en\"=\"Dian\"] (24.388848,102.1697441,26.548485,103.6683522);\n 
#> relation [\"name:en\"=\"Dian\"] (24.388848,102.1697441,26.548485,103.6683522);\n\n);
#> \n(._;>;);\nout body;"
x1 <- osmdata_sf (q1)
length (x1$osm_polygons)
#> [1] 30
```

<sup>Created on 2021-08-23 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0.9000)</sup>

Was easy to fix thanks to your excellent sleuthing and `reprex` construction - great work! If you're so inclined, feel free to check out the above commits and PR any improvements, especially ones that more clearly distinguish the **AND** workflow that pipes multiple `add_osm_feature()` with the **OR** workflow of a compound `add_osm_features()`. Thanks for the help!

**Comment by Robinlovelace** [comment]
Just saw this pop-up on my GitHub alerts and wanted to say: great example of benefits of open issue tracker. Fast fix Mark, inspiring stuff!

**Comment by anthonynorth** [comment]
Thanks for the fast fix @mpadge.

Potential improvement: The help documentation for `add_osm_feature` and `add_osm_features` could briefly explain how these two functions differ, a copy-paste of the explanation in the readme could be enough.

Or something to the effect of "... chained calls to `add_osm_feature()` or `add_osm_features()` have the effect of combining features with a logical AND; this is analogous to chaining `dplyr::filter()` on a data.frame. Features defined in a single `add_osm_features()` call are combined with a logical OR"

**Comment by mpadge** [comment]
Thanks @anthonynorth, it'd be great if you could submit a pull request with those suggested changes!

**Comment by mpadge** [comment]
I'll also re-open this issue until the documentation has been appropriately updated. @anthonynorth If you do that via a PR, please add `"closes #240"` or similar to your final commit message to close this issue. Thanks!
